src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1316,3 +1321,103 @@ uct_dc_mlx5_query_tl_devices(uct_md_h md, uct_tl_device_resource_t **tl_devices_
 UCT_TL_DEFINE(&uct_ib_component, dc_mlx5, uct_dc_mlx5_query_tl_devices,
               uct_dc_mlx5_iface_t, "DC_MLX5_", uct_dc_mlx5_iface_config_table,
               uct_dc_mlx5_iface_config_t);
+
+static void
+uct_dc_mlx5_dci_keepalive_handle_failure(uct_dc_mlx5_iface_t *iface,
+                                         struct mlx5_cqe64 *cqe,
+                                         uint8_t dci,
+                                         ucs_status_t ep_status)
+{
+    uint16_t hw_ci = ntohs(cqe->wqe_counter);
+    uct_rc_txqp_t *txqp;
+    uct_ib_mlx5_txwq_t *txwq;
+    uct_dc_mlx5_ep_t *ep;
+    uct_rc_iface_send_op_t *op;
+    ucs_queue_elem_t *elem;
+
+    ucs_assert(dci == iface->tx.ndci);
+    UCT_DC_MLX5_IFACE_TXQP_DCI_GET(iface, dci, txqp, txwq);
+    elem = ucs_queue_pull(&txqp->outstanding);
+    if (elem == NULL) {
+        /* outstanding list is empty, just exit */
+        goto restart_dci;
+    }
+
+    op = ucs_container_of(elem, uct_rc_iface_send_op_t, queue);
+    if (hw_ci != op->sn) {
+        goto put_op;
+    }
+
+    ep = ucs_derived_of(op->ep, uct_dc_mlx5_ep_t);
+    if (ep->dci != UCT_DC_MLX5_EP_NO_DCI) {
+        uct_dc_mlx5_ep_handle_failure(ep, cqe, ep_status);
```

Comment:
add comment that this ep has posted ops on another dci

Reply:
added comment